Add setup function for pragma, forwardRef, and theme

donysukardi opened the pull request:
**Pending tasks**

- [ ] Tests
- [ ] Documentation
- [ ] Bundle size

This change increases the bundle size though. Any idea how we could trim it to fit the maxSize?

```
 FAIL  ./dist/goober.js: 1.03KB > maxSize 1KB (gzip)

 FAIL  ./dist/goober.module.js: 1.03KB > maxSize 1KB (gzip)
```

Related Issues: Theming (#61), forwardRef (#63)

cristianbote commented on src/styled.js:
```diff
@@ -1,29 +1,37 @@
-        return function Styled(props) {
-            const _props = (_ctx.p = Object.assign({}, props));
+        function Styled(props, ref) {
+            const theme = useTheme();
```
Does it has to be a function? Could it be a plain object? Could become a perf bottleneck on user land.

donysukardi replied:
This is to get feature parity with other CSS-in-JS theming capability, i.e.
```jsx
const useTheme = () => useContext(ThemeContext)
```

Here's how it's similarly done in `styled-components@5` - https://github.com/styled-components/styled-components/blob/v5/packages/styled-components/src/models/StyledComponent.js#L129

donysukardi replied:
For non context user, it could be just `() => theme`

cristianbote commented on src/styled.js:
```diff
@@ -1,29 +1,37 @@
             _props.className =
                 css.apply(_ctx, _args) + (_previousClassName ? ' ' + _previousClassName : '');
+            if (ref) {
+                _props.ref = ref;
```
If we drop the `if` and always do `_props.ref = ref;` would that be strange? 🤔

donysukardi replied:
In terms of functionality, I don't think it'd be any different. As `ref` is treated as special props, it won't showed up in component props anyway.

cristianbote commented:
This is really great! Really love it! How do you feel about it? Can you add the tests and everything? What can I help you with?

cristianbote commented on src/styled.js:
```diff
@@ -1,29 +1,37 @@
-let h;
-const setPragma = pragma => (h = pragma);
+let h, forwardRef, useTheme;
+const setup = (_h, _forwardRef, _useTheme = () => {}) => {
```
How about not defining a default `_useTheme`? Would that break?

cristianbote commented on src/styled.js:
```diff
@@ -1,29 +1,37 @@
-            const _props = (_ctx.p = Object.assign({}, props));
+        function Styled(props, ref) {
+            const theme = useTheme();
+            const _props = (_ctx.p = Object.assign(theme ? { theme } : {}, props));
```
Also what do you think?
```suggestion
            const _props = (_ctx.p = Object.assign(useTheme ? useTheme() : {}, props));
```

donysukardi replied:
Technically it's possible, but usually `useTheme` just returns `theme` instead of `{ theme }`, e.g. https://github.com/styled-components/styled-components/blob/v5/packages/styled-components/src/hooks/useTheme.js

It'd have to be like this instead,
```
const _props = (_ctx.p = Object.assign(useTheme ? { theme: useTheme() } : {}, props));
```

Let me give it a spin and see how the bundle size looks like, coz with gzip more code != bigger size and vice-versa :grin:

cristianbote commented:
Did I mention how exciting this looks? 😄 Can't wait to have the theme support in. I can see myself using it, for sure!

This is a good one! Really appreciate your time on this! 👍

donysukardi commented:
Done the changes for both `useTheme` and `ref`. Bundle size seems to be lower now.

``` 
FAIL  ./dist/goober.js: 1.02KB > maxSize 1KB (gzip)

 FAIL  ./dist/goober.module.js: 1.01KB > maxSize 1KB (gzip)
```

I'll work on the tests for this.

nksaraf commented:
Hi,
I've been following the development of goober and absolutely love the simplicity of the code base. I have had to implement similar functionality for a fork. Is there anything I can do to help land this PR, or in other ways with the library.

cristianbote commented:
Hey @nksaraf thanks for checking goober all along! 🙂 Well anything that comes to mind! There a couple more things inside the issues that can be tackled as well. Give them a try see what you can do.

Thank you for your time!

donysukardi commented:
@cristianbote I've updated the implementation slightly. I observed that in styled-components, the `theme` from context is not passed down to the component, i.e. it's just used to generate the className with necessary styles.

I've also added unit and integration tests for the changes.

ryansolid commented:
This was going strong up to a month or so ago. Is this PR still moving forward? Is something holding it up? Are we still holding out hope it can get 0.02 kb smaller?

cristianbote commented:
Hey @ryansolid! :) Thanks for stopping by!

Well, it's just really my lack of time to do a perf bench on this and check the sizes. Also, this will definitely be a major version upgrade, since the whole `setPragma` update.

cristianbote commented:
Oh and documentation about the changes.

cristianbote commented:
Had a look into this change sets and it seems that this would add some considerable size. Something like `1035B`. Will dig more into it.

JoviDeCroock commented on src/styled.js:
```diff
@@ -1,29 +1,40 @@
+                    css.apply(_ctx, _args) + (_previousClassName ? ' ' + _previousClassName : '')
+            };
 
             return h(tag, _props);
```
```suggestion
            return h(tag, Object.assign({ ref, className: css.apply(_ctx, _args) + (_previousClassName ? ' ' + _previousClassName : '')  }, props));
```

This removes the need for line 26-31

JoviDeCroock commented on src/styled.js:
```diff
@@ -1,29 +1,40 @@
     return function() {
         const _args = arguments;
+        function Styled(props, ref) {
+            _ctx.p = {
```
```suggestion
            _ctx.p = Object.assign({ theme: useTheme ? useTheme() : undefined }, props)
```

cristianbote commented:
Hey @donysukardi things are happening! 🥳😎 Thanks to @JoviDeCroock for pushing this with his PR for autoprefixer. Overall this will end up under 1024B 🥳👍🥜 wohoo! So excited!

This will definitely be v2

cristianbote commented:
The tests failures are my doing 😑 sorry. I resolved the conflicts but somehow missed stuff. Will correct then asap

donysukardi commented:
This is awesome! I am excited!! Let me know if there's anything else I can help with :D

cristianbote commented:
Awesome! @donysukardi I think @JoviDeCroock opened a PR against yours to fix the test failures. In here https://github.com/donysukardi/goober/pull/1/files I think that would make this to pass.

cristianbote commented:
Alrighty! 😄 everything is fine. The above failures are expected 👍 

Thank you soo much @donysukardi! 🎉 🌟
